src/nbl/asset/utils/CGLSLCompiler.cpp

```diff
@@ -0,0 +1,155 @@
+            const shaderc_shader_kind stage = options.stage == IShader::ESS_UNKNOWN ? shaderc_glsl_infer_from_source : ESStoShadercEnum(options.stage);
+            const size_t glsl_len = strlen(code);
+            if (options.genDebugInfo)
+                shadercOptions.SetGenerateDebugInfo();
+
+            shaderc::AssemblyCompilationResult asm_res;
+            shaderc::SpvCompilationResult bin_res;
+            if (options.outAssembly)
+            {
+                asm_res = comp.CompileGlslToSpvAssembly(code, glsl_len, stage, options.sourceIdentifier ? options.sourceIdentifier : "", shadercOptions);
+                options.outAssembly->resize(std::distance(asm_res.cbegin(), asm_res.cend()));
+                memcpy(options.outAssembly->data(), asm_res.cbegin(), options.outAssembly->size());
+                bin_res = comp.AssembleToSpv(options.outAssembly->data(), options.outAssembly->size(), shadercOptions);
+            }
+            else
+            {
+                bin_res = comp.CompileGlslToSpv(code, glsl_len, stage, options.sourceIdentifier ? options.sourceIdentifier : "", options.entryPoint, shadercOptions);
+            }
+
+            if (bin_res.GetCompilationStatus() == shaderc_compilation_status_success)
+            {
+                outSpirv = core::make_smart_refctd_ptr<ICPUBuffer>(std::distance(bin_res.cbegin(), bin_res.cend()) * sizeof(uint32_t));
+                memcpy(outSpirv->getPointer(), bin_res.cbegin(), outSpirv->getSize());
+
+                if (options.spirvOptimizer)
+                    outSpirv = options.spirvOptimizer->optimize(outSpirv.get(), options.logger);
+            }
+            else
+            {
+                options.logger.log(bin_res.GetErrorMessage(), system::ILogger::ELL_ERROR);
+            }
+        }
+        else
+        {
+            options.logger.log("shaderc requires entry point to be \"main\" in GLSL", system::ILogger::ELL_ERROR);
+        }
+    }
+    else
+    {
+        options.logger.log("code is nullptr", system::ILogger::ELL_ERROR);
+    }
+    return outSpirv;
+}
+
+core::smart_refctd_ptr<ICPUShader> CGLSLCompiler::createSPIRVShader(const char* code, const CGLSLCompiler::SOptions& options) const
+{
+    auto spirv = compileToSPIRV(code, options);
+    if (spirv)
+        return core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(spirv), options.stage, IShader::E_CONTENT_TYPE::ECT_GLSL, options.sourceIdentifier);
+    else
+        return nullptr;
+}
+
+core::smart_refctd_ptr<ICPUShader> CGLSLCompiler::createSPIRVShader(system::IFile* sourceFile, const CGLSLCompiler::SOptions& options) const
+{
+    size_t fileSize = sourceFile->getSize();
+    std::string code(fileSize, '\0');
+
+    system::IFile::success_t success;
+    sourceFile->read(success, code.data(), 0, fileSize);
+    if (success)
+        return createSPIRVShader(code.c_str(), options);
+    else
+        return nullptr;
+}
```

Comment:
hmm, could be hoisted to `IShaderCompiler` methinks

Reply:
no because the compile function is not virtual. If it was it had to take `IShaderCompiler::SOptions` instead of `HLSL/GLSLCompiler::SOptions` then we would've needed to do the conversions inside the compile functions instead of `CCompilerSet` meaning get the `getType` of Options and cast them to appropriate struct.

I can make them virtual and do the second option. I just had to make a choice and It was weird to me to do those casts and checks inside the compile functions.

Reply:
the code is identical

you can make the `createSPIRVShader(, const IShaderCompiler::SOptions&)` a protected virtual so no one abuses it :)

And then have the ones that take the actual proper otpion structs non virtual and call the virtual one

Reply:
hmmm, I don't like this. We'll be casting back and forth multiple times. CCompilerSet forces the struct to be `CHLSLCompiler::SOptions` so then It calls `CHLSLCompiler::createSPIRVShader`  --calls-->  `createSPIRVShader(, const IShaderCompiler::SOptions&)` which may cast AGAIN to `CHLSLCompiler::SOptions` inside and do the same checks we do in CCompilerSet.

If you wanna go this route I suggest just having a public virtual `createSPIRVShader(, const IShaderCompiler::SOptions&)` and do checks for type and cast inside the function

Reply:
> If you wanna go this route I suggest just having a public virtual createSPIRVShader(, const IShaderCompiler::SOptions&) and do checks for type and cast inside the function

yes but that opens the door to abuse, which is to pass the wrong option struct.

Reply:
Seem like this is the only way if you want to make this function that takes a file virtual.

but how are you going to prevent people from abusing this `createSPIRVShader(system::IFile* sourceFile, const IShaderCompiler::SOptions& options)`?

When we take a file we take IShaderCompiler::SOptions in the virtual function but when we want to give the source directly we have to use CHLSL/GLSLCompiler::SOptions in the non-virtual functions. that's just inconsistent